File: 20/python/aoc2.py

def is_valid(char_range, char, passwd):
    range_split = char_range.split("-")
    char_count = 0
    for c in passwd:
        if c == char:
            char_count += 1
    return char_count >= int(range_split[0]) and char_count <= int(range_split[1])

def xor(x, y):
    return bool(x)+bool(y) == 1

def is_valid2(char_range, char, passwd):
    range_split = char_range.split("-")
    return xor(passwd[int(range_split[0]) - 1] == char,  passwd[int(range_split[1]) - 1] == char)

def valid_passwords(arr):
    valid = 0
    for line in arr:
        split = line.split(" ")
        char_range = split[0]
        char = split[1][0]
        passwd = split[2]
        if is_valid2(char_range, char, passwd):
            valid += 1
    return valid


if __name__ == "__main__":
    text_file = open("../input/aoc2.txt", "r")
    string_data = text_file.readlines()
    ans = valid_passwords(string_data)
    print("the number of valid passwords is {}".format(ans))
   

File: 20/go/aoc2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func main() {
	data, err := ReadFile("../input/aoc2.txt")
	if err != nil {
		fmt.Printf("readFile failed: %v\n", err)
	}
	ans := validPasswords(data)
	fmt.Printf("There are %d valid passwords\n", ans)
}

func validPasswords(data []string) int {
	valid := 0
	// example line -> 1-3 a: abcde
	for _, line := range data {
		if len(line) < 1 {
			continue
		}
		split := strings.Split(line, " ")
		password := split[2]
		char := split[1][0]
		countRange := split[0]
		if isPasswordValid2(password, char, countRange) {
			valid++
		}
	}
	return valid
}

func isPasswordValid(pass string, targetChar byte, countRange string) bool {
	m := make(map[byte]int)
	bs := []byte(pass)
	// populate map count
	for _, chr := range bs {
		if val, ok := m[chr]; ok {
			m[chr] = val + 1
		} else {
			m[chr] = 1
		}
	}
	targetCount, _ := m[targetChar]
	splitRange := strings.Split(countRange, "-")
	low, err := strconv.Atoi(splitRange[0])
	if err != nil {
		fmt.Printf("There was an error converting the low range in %s \n", countRange)
	}
	high, err := strconv.Atoi(splitRange[1])
	if err != nil {
		fmt.Printf("There was an error converting the high range in %s \n", countRange)
	}

	if targetCount <= high && targetCount >= low {
		return true
	}
	return false

}

func isPasswordValid2(pass string, targetChar byte, countRange string) bool {
	splitRange := strings.Split(countRange, "-")
	low, err := strconv.Atoi(splitRange[0])
	if err != nil {
		fmt.Printf("There was an error converting the low range in %s \n", countRange)
	}
	high, err := strconv.Atoi(splitRange[1])
	if err != nil {
		fmt.Printf("There was an error converting the high range in %s \n", countRange)
	}
	// XOR
	if (pass[low-1] == targetChar) != (pass[high-1] == targetChar) {
		return true
	}
	return false

}
